refactor: Update controller packages to v44

Gudahtt opened the pull request:
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

Controller packages have been updated to roughly match the versions used in v44 of the core monorepo (this doesn't include the keyring controller, and a few packages are ahead of v44 already).

This update includes a significant breaking change: the migration of `frequentRpcList` in the `PreferencesController` to
`networkConfigurations` in the `NetworkController`. Various other packages have also been updated at the same time because they have a peer dependency upon the `NetworkController`.

**Issue**

Closes https://github.com/MetaMask/mobile-planning/issues/798

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

Gudahtt committed:
Update controller packages to core v44

Controller packages have been updated to roughly match the versions
used in v44 of the core monorepo (this doesn't include the keyring
controller, and a few packages are ahead of v44 already).

This update includes a significant breaking change: the migration of
`frequentRpcList` in the `PreferencesController` to
`networkConfigurations` in the `NetworkController`. Various other
packages have also been updated at the same time because they have a
peer dependency upon the `NetworkController`.

Closes https://github.com/MetaMask/mobile-planning/issues/798

Gudahtt commented:
@SocketSecurity ignore @metamask/swappable-obj-proxy@2.1.0

This is our package

Gudahtt commented:
This depends upon:

- https://github.com/MetaMask/metamask-mobile/pull/6874
- https://github.com/MetaMask/metamask-mobile/pull/6878
- https://github.com/MetaMask/metamask-mobile/pull/6879
- https://github.com/MetaMask/metamask-mobile/pull/6880
- https://github.com/MetaMask/metamask-mobile/pull/6881
- https://github.com/MetaMask/metamask-mobile/pull/6884

Gudahtt commented on app/core/BackgroundBridge/BackgroundBridge.js:
```diff
@@ -59,15 +55,6 @@ export class BackgroundBridge extends EventEmitter {
-    const provider = Engine.context.NetworkController.provider;
-    const blockTracker = provider._blockTracker;
-
-    // provider and block tracker proxies - because the network changes
```
The `NetworkController` method `getProviderAndBlockTracker` now returns a proxied provider and block tracker, so we don't need to proxy them at this layer anymore

Gudahtt commented on package.json:
```diff
@@ -154,24 +154,24 @@
     "@metamask/address-book-controller": "^2.0.0",
     "@metamask/approval-controller": "^3.4.0",
-    "@metamask/assets-controllers": "5.0.0",
+    "@metamask/assets-controllers": "~5.0.1",
```
This is kept at v5.0.1 because v5.1.0 depends upon v7 of the network controller

Gudahtt commented:
Successful e2e test run: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/57d8e065-650f-4e01-9fb8-577cb754b96f

cortisiko commented:
@Gudahtt Once a successful dev review is done, let's put this PR through QA. I am a bit nervous about only having e2e run against it.

Gudahtt commented:
Definitely!

Gudahtt commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -542,29 +542,39 @@ class NetworkSettings extends PureComponent {
         },
       );
-      // TODO: PreferencesController.addToFrequentRpcList only compares RPC urls to determine if a network should be updated or added.
+      // TODO: Use network configuration ID to update existing entries
```
Leaving it in this state is far from ideal. If we used the network configuration ID instead, we could eliminate a lot of this duplicate detection logic.

However, I started going down this path and lost confidence in it. This component is pretty complicated and I didn't understand all of the moving parts well enough. Instead I left it to use the same strategy it did before, because it seemed less likely to break something.

mcmire commented on app/components/UI/Ramp/components/OrderDetails.tsx:
```diff
@@ -176,15 +176,15 @@ interface Props {
+   * Network configurations
    */
-  frequentRpcList: any;
+  networkConfigurations: any;
```
Should we use `NetworkState['networkConfigurations']` here?

Gudahtt replied:
Oops, wrong line. Fixed here instead: 303697265

mcmire commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -303,16 +305,16 @@ class NetworkSettings extends PureComponent {
-    const { frequentRpcList } = this.props;
-    const networkInformation = frequentRpcList.find(
+    const { networkConfigurations } = this.props;
+    const networkInformation = Object.values(networkConfigurations).find(
       ({ chainId: id }) => String(id) === String(Networks.mainnet.chainId),
     );
     return networkInformation?.rpcUrl || '';
```
Nit: This now has a proper name (`networkConfiguration`).

``` suggestion
    const networkConfiguration = Object.values(networkConfigurations).find(
      ({ chainId: id }) => String(id) === String(Networks.mainnet.chainId),
    );
    return networkConfiguration?.rpcUrl || '';
```

Gudahtt replied:
Done here: 7104b9bd3

mcmire commented on app/components/Views/Settings/NetworksSettings/NetworkSettings/index.js:
```diff
@@ -337,7 +339,7 @@ class NetworkSettings extends PureComponent {
         }
       } else {
-        const networkInformation = frequentRpcList.find(
+        const networkInformation = Object.values(networkConfigurations).find(
```
Nit: This now has a proper name (`networkConfiguration`).

Gudahtt replied:
Done here: 7104b9bd3

mcmire commented on app/components/Views/Settings/NetworksSettings/index.js:
```diff
@@ -283,20 +296,22 @@ class NetworksSettings extends PureComponent {
+    const { networkConfigurations } = this.props;
+    return Object.values(networkConfigurations).map(
+      ({ rpcUrl, nickname, chainId }, i) => {
+        const { name } = { name: nickname || rpcUrl };
```
Nit: This can be simplified to

``` suggestion
        const name = nickname || rpcUrl;
```

mcmire commented on app/core/RPCMethods/wallet_switchEthereumChain.js:
```diff
@@ -55,34 +54,37 @@ const wallet_switchEthereumChain = async ({
       return;
     }
 
+    const [networkConfigurationId, networkConfiguration] = existingEntry;
```
It seems like `existingEntry` could be undefined here (if `existingNetworkDefault` is not nullish instead). Do we want to check for that here somehow before destructuring?

mcmire reviewed (commented):
Did a first pass, left some minor comments. I will now check the changelog for v44 and make sure all changes are accounted for.

mcmire reviewed (dismissed):
I have double-checked the changes in these packages to ensure that they are covered by these changes. I am not familiar with all of the areas of mobile but the changes here make sense to me.

Gudahtt committed:
Improve prop type

Gudahtt committed:
Improve local variable names

Gudahtt committed:
Simplify variable assignment

Co-authored-by: Elliot Winkler <elliot.winkler@gmail.com>

Cal-L reviewed (commented):
Looks like the `chainId` that's being stored in the network controller for mobile is decimal format. When passed into `NetworkController.upsertNetworkConfiguration`, the app crashes with the error `Value must be a hexadecimal string, starting with "0x"`. It seems like we'll have to update the entire app to use that format but probably doesn't make sense to do it in this PR. We can just convert the decimal chain ID -> hexadecimal before passing it into `upsertNetworkConfiguration` for now.

Cal-L reviewed (commented):
Other than the `upsertNetworkConfiguration` chainId issue, everything else including migrations LGTM

Gudahtt committed:
Add patch to remove chain ID hex validation

The method `upsertNetworkController` has validation to ensure the chain
ID is a hex string. This was introduced here prematurely, before the
core packages were updated to use hex instead of decimal. This
validation has been removed with a patch, which we can then remove when
updating to the core v53 packages (which updates chain ID to be hex in
every package).

Gudahtt commented:
Great catch @Cal-L ! This was a problem I was aware of but had forgotten about - hex validation was added to that method despite the fact that we still use decimal string chain IDs in that controller. If we converted it to hex for just that call, it would remain in the wrong format in the configuration and would cause further problems. Instead I have updated `upsertNetworkConfiguration` in a patch to expect a decimal string chain ID, until the v53 update which converts it to hex.

Gudahtt committed:
Improve type used for network configurations prop

Gudahtt commented:
Updated passing e2e test run here: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/c68e73a6-29b8-45fe-85f8-f9240c2fb938

Cal-L reviewed (dismissed):
Confirmed the network switch issue is resolved

https://github.com/MetaMask/metamask-mobile/assets/10508597/8621c188-8cce-4709-87c0-6ad01eaf28f2

LGTM

Gudahtt commented:
Rebased to resolve conflicts

cortisiko reviewed (changes requested):
@Gudahtt I found a bug while testing. 

When adding a custom network either by adding it through the popular networks flow, or manually adding a custom network, the network does not automatically switch from mainnet to the custom network by default. 

To reproduce:
Import a wallet or create a wallet
Go to through the add custom network flow (tap on the network picker in the top nav on the wallet view > Add network > custom network tab)
Add a custom network using an RPC URL. Feel free to use gnosis: RPC URL: 

Notice you do not switch from mainnet to the custom network: http://recordit.co/9De0nA7ocy  

Reproducing via the custom network flow:
Import a wallet or create a wallet
Go to through the add custom network flow (tap on the network picker in the top nav on the wallet view > Add network > popular network tab)
Approve to add the network to your network list 
When the network-added modal appears tap the switch to network
Notice you are not taken to the custom network 

See here: http://recordit.co/v5iWMAR6GR 

Expected behavior:
When adding a custom network manually the network should automatically switch from mainnet to the custom network by default. 
 http://recordit.co/pIbty7aWDG 

Similarly, if you were to add a network via the popular networks flow, the network does not automatically update.

Gudahtt committed:
Fix automatic network switching

The network should switch automatically when adding a custom network
manually. This was broken because of an invalid option (`active` was
used instead of `setActive`).

Gudahtt committed:
Add comment to patch explaining how it was generated and what it is for

Gudahtt commented:
The reported bug should hopefully be fixed! Though I haven't validated that yet.

Updated Bitrise e2e smoke test run: https://app.bitrise.io/app/be69d4368ee7e86d/pipelines/27637e9f-c16d-4d77-b6ff-b3e8ca57d3a6

Bitrise browserstack test run: https://app.bitrise.io/build/fad7d051-9bb5-4eb9-bf9f-a407cdfed4da

Both passing

cortisiko reviewed (approved):
Fix looks good! Nothing else to report 🌮
